### PR TITLE
Add test for CWalletTx::GetImmatureCredit() returning stale values.

### PR DESCRIPTION
**Pull request by ryanofsky**
~~Fix CWalletTx::GetImmatureCredit() returning stale values.~~

~~One line fix. Add missing assignment to CWalletTx::MarkDirty() to clear the cached immature credit flag.~~

Updated description: Add test for cached immature credit flag not being cleared in CWalletTx::MarkDirty() bug, which was fixed in #8717, commit a560378.

### CODE REVIEW

**Comment by laanwj** [comment]
Seems a duplicate of #8717?

**Comment by ryanofsky** [comment]
@laanwj, it's basically the same as #8717 (which I didn't know about) except it adds the test and places the new assignment in between fCreditCached and fAvailableCreditCached following the order of CWalletTx class members. So I think it's a slightly better PR.

**Comment by laanwj** [comment]
Right. Will merge #8717 first, then we can rebase this into a test-only pull.

File: src/wallet/test/wallet_tests.cpp [hunk 1]
```diff
@@ -351,4 +353,28 @@ BOOST_AUTO_TEST_CASE(ApproximateBestSubset)
...
+BOOST_FIXTURE_TEST_CASE(coin_mark_dirty_immature_credit, TestChain100Setup)
+{
+    CWallet wallet;
+    CWalletTx wtx(&wallet, MakeTransactionRef(coinbaseTxns.back()));
+    wtx.hashBlock = chainActive.Tip()->GetBlockHash();
+    wtx.nIndex = 0;
+
+    // Call GetImmatureCredit() once before adding the key to the wallet to
+    // cache the current immature credit amount, which is 0.
+    BOOST_CHECK_EQUAL(wtx.GetImmatureCredit(), 0);
```
Comment (MarcoFalke) [inline src/wallet/test/wallet_tests.cpp]: Travis:

```
Assertion failed: lock cs_main not held in ../../src/wallet/wallet.cpp:3761; locks held:
Running 221 test cases...
unknown location(0): fatal error: in "wallet_tests/coin_mark_dirty_immature_credit": signal: SIGABRT (application abort requested)
../../src/wallet/test/wallet_tests.cpp(371): last checkpoint
test_bitcoin: ../../src/key.cpp:300: void ECC_Start(): Assertion `secp256k1_context_sign == __null' failed.
unknown location(0): fatal error: in "wallet_tests/coin_selection_tests": signal: SIGABRT (application abort requested)
../../src/wallet/test/wallet_tests.cpp(70): last checkpoint: "coin_selection_tests" fixture entry.
test_bitcoin: ../../src/key.cpp:300: void ECC_Start(): Assertion `secp256k1_context_sign == __null' failed.
unknown location(0): fatal error: in "wallet_crypto/decrypt": signal: SIGABRT (application abort requested)
../../src/wallet/test/crypto_tests.cpp(217): last checkpoint: "decrypt" fixture entry.

Comment (ryanofsky) [inline src/wallet/test/wallet_tests.cpp]: Thanks, the locking error has been fixed.

**Comment by kallewoof** [comment]
ACK 52045d1 (test passes on version rebased onto master)

**Comment by ryanofsky** [comment]
Rebased 52045d1bb71cd4be7ae269c63a8260df02a8cb56 -> 7ed143c10ec7c34843cb025951a6bbf279d1774d ([immature.5](https://github.com/ryanofsky/bitcoin/commits/pr/immature.5) -> [immature.6](https://github.com/ryanofsky/bitcoin/commits/pr/immature.6)) because of conflict with recent changes in wallet_tests.cpp.

**Comment by MarcoFalke** [comment]
utACK 7ed143c10ec7c34843cb025951a6bbf279d1774d. Sorry, somehow forgot about this pull.

**Pull request merged**